### PR TITLE
fix(Observer): observer interface now allows partial Observers

### PR DESCRIPTION
**Pull request by blesh**
- caveat is that there is no way to require \_at least one\_ method on Observer in TypeScript, at least that I can find

related #1260

### CODE REVIEW

**Comment by kwonoj** [comment]
Looks good to me for now. I'll track how typescript side changes / updates are going to followup further.

**Comment by kwonoj** [comment]
Seems with current Typescript it can be achieved via below, not looking elegant bit though..

`type Observer<T> = NextObserver<T> | ErrorObserver | CompleteObserver | (NextObserver<T> & ErrorObserver & CompleteObserver) | (ErrorObserver & CompleteObserver) | (NextObserver<T> & ErrorObserver) | (NextObserver<T> & CompleteObserver);`

(copy-pasted from issue).

**Comment by blesh** [comment]
I experimented for a while with different approaches and this was the best I could come up with for developer ergonomics.

I tried this also: 

```TypeScript

export interface NextObserver<T> {
  isUnsubscribed?: boolean;
  next: (value: T) => void;
}

export interface ErrorObserver {
  isUnsubscribed?: boolean;
  error: (err: any) => void;
}

export interface CompletionObserver {
  isUnsubscribed?: boolean;
  complete: () => void;
}

export declare type Observer<T> = NextObserver<T> | ErrorObserver | CompletionObserver;
```

But that resulted in ugly experiences on the consumer end, where trying to call `next` on the Observer gave you TypeScript errors because "next" doesn't exist on " NextObserver<T> | ErrorObserver | CompletionObserver"

**Comment by blesh** [comment]
@kwonoj ... haha you typed that while I was adding additional comments. Yes, I did try that, but it ended up being ugly on the side of whatever _consumes_ the observer. For example and error like: `o.next does not exist on NextObserver<T> | ErrorObserver | CompleteObserver | (NextObserver<T> & ErrorObserver & CompleteObserver) | (ErrorObserver & CompleteObserver) | (NextObserver<T> & ErrorObserver) | (NextObserver<T> & CompleteObserver)`;

**Comment by kwonoj** [comment]
Yes, it'll be quite unfriendly from consumer perspective. For now I'd rather allow observer to be fully optional interfaces given that most cases caller wouldn't pass empty object ({}) in usual cases. Course this is based on assumption there'll be better ergonomics support from typescript will be available. I hope https://github.com/Microsoft/TypeScript/issues/4889 can be answer.

**Comment by blesh** [comment]
@kwonoj ... I got a solid answer from the TypeScript folks and reimplemented my solution. Now there is a `PartialObserver<T>` type that matches any partial observer. I've also updated all appropriate locations in code that could have partial observers.

**Comment by kwonoj** [comment]
Saw update from issue and this looks much better :) Change looks good to me. 

p.s: just curious, how does error output now when partial observer does not have some interfaces? (as compare to above)

**Comment by blesh** [comment]
@kwonoj if you try to pass an empty object, for example to something that accepts `PartialObserver<T>`, it complains that `{} does not match NextObserver<T> | ErrorObserver | CompletionObserver`... likewise if there's too many additional methods: `{ next() {}, blah () {} }`.

**Comment by kwonoj** [comment]
Thanks for clarification :)

File: src/Notification.ts [hunk 1]
```diff
@@ -23,17 +23,17 @@ export class Notification<T> {
...
-        return complete();
+        return complete && complete();
     }
   }
 
-  accept(nextOrObserver: Observer<T> | ((value: T) => void), error?: (err: any) => void, complete?: () => void) {
-    if (nextOrObserver && typeof (<Observer<T>>nextOrObserver).next === 'function') {
-      return this.observe(<Observer<T>>nextOrObserver);
+  accept(nextOrObserver: PartialObserver<T> | ((value: T) => void), error?: (err: any) => void, complete?: () => void) {
+    if (nextOrObserver && typeof (<PartialObserver<T>>nextOrObserver).next === 'function') {
```
Comment (saneyuki) [inline src/Notification.ts]: By the way, I think this guard will accepts `IterableIterator<T>` (from the code which have only a runtime dynamically type). In today's RxJS context,  can a generator be a `Observer<T>`?

Comment (blesh) [inline src/Notification.ts]: @saneyuki ... not really... generators have `next`, `throw` and `return` methods. Very, very early versions of this library where implemented in that manner, but it was abandoned when the es-observable spec moved away from that.

**Pull request merged**